InstantSearch.js/segment-connector/src/app.js

```diff
@@ -0,0 +1,128 @@
+/* global algoliasearch analytics */
+
+import instantsearch from 'instantsearch.js';
+import {
+  searchBox,
+  hits,
+  refinementList,
+  pagination,
+  rangeSlider,
+} from 'instantsearch.js/es/widgets';
+import insightsClient from 'search-insights';
+import { createInsightsMiddleware } from 'instantsearch.js/es/middlewares';
+
+const searchClient = algoliasearch(
+  'latency',
+  '6be0576ff61c053d5f9a3225e2a90f76'
+);
+
+const search = instantsearch({
+  indexName: 'instant_search',
+  searchClient,
+});
+
+search.addWidgets([
+  searchBox({
+    container: '#searchbox',
+  }),
+  hits({
+    container: '#hits',
+    templates: {
+      item: (hit, bindEvent) => {
+        const result = `
+        <article>
+          <h1>${instantsearch.highlight({ attribute: 'name', hit })}</h1>
+          <p>${instantsearch.highlight({ attribute: 'description', hit })}</p>
+          <button type="button" ${bindEvent('view', hit, 'View')}>
+            View Product
+          </button>
+          <button type="button" ${bindEvent('click', hit, 'Favorite')}>
+            Favorite
+          </button>
+          <button type="button" ${bindEvent('conversion', hit, 'Add to Cart')}>
+            Add to Cart
+          </button>
+          <button type="button" ${bindEvent('conversion', hit, 'Order')}>
+            Order Now
+          </button>
+        </article>
+      `;
+        return result;
+      },
+    },
+  }),
+  refinementList({
+    container: '#brand-list',
+    attribute: 'brand',
+  }),
+  rangeSlider({
+    container: '#price',
+    attribute: 'price',
+  }),
+  pagination({
+    container: '#pagination',
+  }),
+]);
+
+const insightsMiddleware = createInsightsMiddleware({
+  insightsClient,
+  onEvent(event) {
+    const { widgetType, eventType, payload, hits, attribute } = event;
+
+    console.log(event);
+    if (widgetType === 'ais.hits' && eventType === 'view') {
+      analytics.track('Product List Viewed', {
+        index: payload.index,
+        products: hits.map((hit) => ({
+          objectID: hit.objectID,
```

Comment:
if we only needed the objectIDs, could it not have been read from the payload instead?

Reply:
yes, but according to the [Segment spec](https://segment.com/docs/connections/spec/ecommerce/v2/#product-list-viewed), they will for sure put other properties, but I didn't to go too specific, so only put the necessary properties like that as a skeleton.